dex/elrond_dex_farm/src/farm_token.rs

```diff
@@ -0,0 +1,187 @@
+            AsyncCallResult::Ok(token_id) => {
+                self.last_error_message().clear();
+
+                if self.farm_token_id().is_empty() {
+                    self.farm_token_id().set(&token_id);
+                    self.nft_deposit_accepted_token_ids().insert(token_id);
+                }
+            }
+            AsyncCallResult::Err(message) => {
+                self.last_error_message().set(&message.err_msg);
+
+                let (returned_tokens, token_id) = self.call_value().payment_token_pair();
+                if token_id.is_egld() && returned_tokens > 0 {
+                    let _ = self.send().direct_egld(caller, &returned_tokens, &[]);
+                }
+            }
+        }
+    }
+
+    #[endpoint(setLocalRolesFarmToken)]
+    fn set_local_roles_farm_token(&self) -> SCResult<AsyncCall<Self::SendApi>> {
+        require!(self.is_active(), "Not active");
+        self.require_permissions()?;
+        require!(!self.farm_token_id().is_empty(), "No farm token issued");
+
+        let token = self.farm_token_id().get();
+        Ok(self.set_local_roles(token))
+    }
+
+    fn set_local_roles(&self, token: TokenIdentifier) -> AsyncCall<Self::SendApi> {
+        ESDTSystemSmartContractProxy::new_proxy_obj(self.send())
+            .set_special_roles(
+                &self.blockchain().get_sc_address(),
+                &token,
+                &[
+                    EsdtLocalRole::NftCreate,
+                    EsdtLocalRole::NftAddQuantity,
+                    EsdtLocalRole::NftBurn,
+                ],
+            )
+            .async_call()
+            .with_callback(self.callbacks().change_roles_callback())
+    }
+
+    #[callback]
+    fn change_roles_callback(&self, #[call_result] result: AsyncCallResult<()>) {
+        match result {
+            AsyncCallResult::Ok(()) => {
+                self.last_error_message().clear();
+            }
+            AsyncCallResult::Err(message) => {
+                self.last_error_message().set(&message.err_msg);
+            }
+        }
+    }
+
+    fn decode_attributes(
+        &self,
+        attributes_raw: &BoxedBytes,
+    ) -> SCResult<FarmTokenAttributes<Self::BigUint>> {
+        let attributes =
+            <FarmTokenAttributes<Self::BigUint>>::top_decode(attributes_raw.as_slice());
+        match attributes {
+            Result::Ok(decoded_obj) => Ok(decoded_obj),
+            Result::Err(_) => {
+                return sc_error!("Decoding error");
+            }
+        }
+    }
+
+    fn get_farm_attributes(
+        &self,
+        token_id: &TokenIdentifier,
+        token_nonce: u64,
+    ) -> SCResult<FarmTokenAttributes<Self::BigUint>> {
+        let token_info = self.blockchain().get_esdt_token_data(
+            &self.blockchain().get_sc_address(),
+            token_id,
+            token_nonce,
+        );
+
+        let farm_attributes = token_info.decode_attributes::<FarmTokenAttributes<Self::BigUint>>();
```

Comment:
can we use `?` here to simplify the error handling?

Reply:
Yes. Will fix

Reply:
Actually we can't. Seems weird but I did some debugging. Here's the reason: https://github.com/ElrondNetwork/elrond-wasm-rs/blob/master/elrond-wasm/src/types/io/sc_error.rs#L57. This has a method for EncodeError but does not have one for DecodeError sadly. We'll stick to the handling until this will be added in the rust framework.

Reply:
ok.